File: Python/TMG39931.py
# Distributed with a free-will license.
# Use it any way you want, profit or free, provided it fits in the licenses of its associated works.
# TMG39931
# This code is designed to work with the TMG39931_I2CS I2C Mini Module available from ControlEverything.com.
# https://www.controleverything.com/content/Color?sku=TMG39931_I2CS#tabs-0-product_tabset-2

import smbus
import time

# Get I2C bus
bus = smbus.SMBus(1)

# TMG39931 address, 0x39(57)
# Select Enable register, 0x80(128)
#		0x0F(15)	Power ON, ALS enable, Proximity enable, Wait disable
bus.write_byte_data(0x39, 0x80, 0x0F)
# TMG39931 address, 0x39(57)
# Select ADC integration time register, 0x81(129)
#		0xFF(255)	ATIME : 712ms, Max count = 65535 cycles
bus.write_byte_data(0x39, 0x81, 0xFF)
# TMG39931 address, 0x39(57)
# Select Wait time register, 0x83(131)
#		0xFF(255)	WTIME : 2.78ms
bus.write_byte_data(0x39, 0x83, 0xFF)
# TMG39931 address, 0x39(57)
# Select Control register, 0x8F(143)
#		0x00(00)	AGAIN is 1x
bus.write_byte_data(0x39, 0x8F, 0x00)

time.sleep(0.8)

# TMG39931 address, 0x39(57)
# Read data back from 0x94(148), 9 bytes
# cData LSB, cData MSB, red LSB, red MSB, green LSB, green MSB, blue LSB, blue MSB, proximity
data = bus.read_i2c_block_data(0x39, 0x94, 9)

# Convert the data
cData = data[1] * 256.0 + data[0]
red = data[3] * 256.0 + data[2]
green = data[5] * 256.0 + data[4]
blue = data[7] * 256.0 + data[6]
proximity = data[8]

# Output data to screen
print "InfraRed luminance : %.2f lux" %cData
print "Red Color luminance : %.2f lux" %red
print "Green Color luminance : %.2f lux" %green
print "Blue Color luminance : %.2f lux" %blue
print "Proximity of the device : %d " %proximity

File: C/TMG39931.c
// Distributed with a free-will license.
// Use it any way you want, profit or free, provided it fits in the licenses of its associated works.
// TMG39931
// This code is designed to work with the TMG39931_I2CS I2C Mini Module available from ControlEverything.com.
// https://www.controleverything.com/content/Color?sku=TMG39931_I2CS#tabs-0-product_tabset-2

#include <stdio.h>
#include <stdlib.h>
#include <linux/i2c-dev.h>
#include <sys/ioctl.h>
#include <fcntl.h>

void main() 
{
	// Create I2C bus
	int file;
	char *bus = "/dev/i2c-1";
	if((file = open(bus, O_RDWR)) < 0) 
	{
		printf("Failed to open the bus. \n");
		exit(1);
	}
	// Get I2C device, TMG39931 I2C address is 0x39(57)
	ioctl(file, I2C_SLAVE, 0x39);

	// Set Wait Time register(0x83)
	// Wtime = 2.78 ms,(0xFF)
	char config[2] = {0};
	config[0] = 0x83;
	config[1] = 0xFF;
	write(file, config, 2);
	// Set Atime register(0x81)
	// ATIME : 712ms, Max count = 65535 cycles
	config[0] = 0x81;
	config[1] = 0x00;
	write(file, config, 2);
	// Select enable register(0x80)
	// Power ON, ALS enable, Proximity enable, Wait enable(0x0F)
	config[0] = 0x80;
	config[1] = 0x0F;
	write(file, config, 2);
	sleep(1);

	// Read 9 Bytes of Data
	// cData lsb, cData msb, red lsb, red msb, green lsb, green msb, blue lsb, blue msb, proximity
	char reg[1] = {0x94};
	write(file, reg, 1);
	char data[9] = {0};
	if(read(file, data, 9) != 9)
	{
		printf("Erorr : Input/output Erorr \n");
	}
	else
	{
		// Convert the data
		int cData = (data[1] * 256 + data[0]);
		int red = (data[3] * 256 + data[2]);
		int green = (data[5] * 256 + data[4]); 
		int blue = (data[7] * 256 + data[6]);
		int proximity = data[8];

		// Output data to screen
		printf("InfraRed luminance : %d lux \n", cData);
		printf("Red Color luminance : %d lux \n", red);
		printf("Green Color luminance : %d lux \n", green);
		printf("Blue Color luminance : %d lux \n", blue);
		printf("Proximity of the device : %d \n", proximity);
	}
}
